src/connector/src/sink/coordinate.rs

```diff
@@ -0,0 +1,152 @@
+                        vnode_bitmap: Some(vnode_bitmap.to_protobuf()),
+                        param: Some(param.to_proto()),
+                    },
+                )),
+            },
+            |req_stream| async move { client.coordinate(req_stream).await },
+        )
+        .await?;
+        match first_response {
+            SinkCoordinatorToWriterMsg {
+                msg: Some(sink_coordinator_to_writer_msg::Msg::StartResponse(_)),
+            } => Ok(Self(stream_handle)),
+            msg => Err(SinkError::Coordinator(anyhow!(
+                "should get start response but get {:?}",
+                msg
+            ))),
+        }
+    }
+
+    async fn commit(&mut self, epoch: u64, metadata: SinkMetadata) -> Result<()> {
+        self.0
+            .send_request(SinkWriterToCoordinatorMsg {
+                msg: Some(sink_writer_to_coordinator_msg::Msg::CommitRequest(
+                    CommitRequest {
+                        epoch,
+                        metadata: Some(metadata),
+                    },
+                )),
+            })
+            .await?;
+        match self.0.next_response().await? {
+            SinkCoordinatorToWriterMsg {
+                msg: Some(sink_coordinator_to_writer_msg::Msg::CommitResponse(_)),
+            } => Ok(()),
+            msg => Err(SinkError::Coordinator(anyhow!(
+                "should get commit response but get {:?}",
+                msg
+            ))),
+        }
+    }
+}
+
+pub struct CoordinatedSinkWriter<W: SinkWriter<CommitMetadata = Option<SinkMetadata>>> {
+    epoch: u64,
+    coordinator_stream_handle: CoordinatorStreamHandle,
+    inner: W,
+}
+
+impl<W: SinkWriter<CommitMetadata = Option<SinkMetadata>>> CoordinatedSinkWriter<W> {
+    pub async fn new(
+        client: SinkCoordinationRpcClient,
+        param: SinkParam,
+        vnode_bitmap: Bitmap,
+        inner: W,
+    ) -> Result<Self> {
+        Ok(Self {
+            epoch: 0,
+            coordinator_stream_handle: CoordinatorStreamHandle::new(client, param, vnode_bitmap)
+                .await?,
+            inner,
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl<W: SinkWriter<CommitMetadata = Option<SinkMetadata>>> SinkWriter for CoordinatedSinkWriter<W> {
+    async fn begin_epoch(&mut self, epoch: u64) -> Result<()> {
+        self.epoch = epoch;
+        self.inner.begin_epoch(epoch).await
+    }
+
+    async fn write_batch(&mut self, chunk: StreamChunk) -> Result<()> {
+        self.inner.write_batch(chunk).await
+    }
+
+    async fn barrier(&mut self, is_checkpoint: bool) -> Result<Self::CommitMetadata> {
+        let start_time = Instant::now();
+        let metadata = self.inner.barrier(is_checkpoint).await?;
+        if is_checkpoint {
+            debug!("take {:?} to fetch metadata", start_time.elapsed());
+        }
+        if is_checkpoint {
+            let metadata = metadata.ok_or(SinkError::Coordinator(anyhow!(
```

Comment:
By design, we want our sink to be aware of barriers because barrier is a boundary for consistency, so the sink framework should call it on non-cp barrier. 

For the interface enforcement, we can divide the current `barrier` method into two methods `barrier` and `checkpoint`, but I'm afraid that will make the trait too verbose.

Reply:
Make sense. I will modify the trait in following PR.